doc/pxe.md

```diff
@@ -0,0 +1,111 @@
+subdirectory) to customize them. In the example below we are just copying all of them.
+
+    zypper in tftpboot-installation-openSUSE-Tumbleweed-x86_64
+    mkdir /srv/tftpboot
+    cp -a /usr/share/tftpboot-installation/openSUSE-Tumbleweed-x86_64 /srv/tftpboot
+
+To define a boot option to run D-Installer, edit the `net/pxelinux.cfg/default` to include:
+
+```
+label iguana
+  ipappend 2
+  kernel boot/x86_64/loader/iguana-new/vmlinuz-iguana
+  append initrd=boot/x86_64/loader/initrd-iguana rd.iguana.control_url=tftp://192.168.122.1/net/d-installer.yaml rd.iguana.debug=1 rd.neednet=1
+```
+
+Do not worry about the kernel, the initrd or the `d-installer.yaml` file, we will jump into it
+later.
+
+## Configure libvirt to server TFTP files
+
+To instruct libvirt to serve the TFTP files, you must add the `tftp` and `bootp` elements to the
+network configuration. Use the `virsh net-edit default` command to edit the configuration and adapt
+it accordingly. Here is an example:
+
+```xml
+<network connections='1'>
+  <name>default</name>
+  <uuid>639e02a7-fcbd-4cf3-a563-6db083aef051</uuid>
+  <forward mode='nat'>
+    <nat>
+      <port start='1024' end='65535'/>
+    </nat>
+  </forward>
+  <bridge name='virbr0' stp='on' delay='0'/>
+  <mac address='52:54:00:fb:7c:8e'/>
+  <ip address='192.168.122.1' netmask='255.255.255.0'>
+    <tftp root='/srv/tftpboot/openSUSE-Tumbleweed-x86_64'/>
+    <dhcp>
+      <range start='192.168.122.2' end='192.168.122.254'/>
+      <bootp file='net/pxelinux.0'/>
+    </dhcp>
+  </ip>
+</network>
+```
+
+## initrd preparation
+
+**Please, build the initrd on a virtual machine to avoid messing up your system.**
+
+Iguana provides a universal initrd in which actual functionality is implemented in containers. Which
+containers to use and how to set them up is defined in a *workflow definition*. The [Iguana
+repository](https://github.com/openSUSE/iguana) includes a [definition for
+D-Installer](https://github.com/openSUSE/iguana/blob/main/iguana-workflow/examples/d-installer.yaml).
+
+In the future, you should be able to get the kernel and the initrd files from the [Iguana
+package](https://build.opensuse.org/package/show/home:oholecek:iguana/iguana). However, for
+D-Installer to work, we need to do include an `/etc/NetworkManager` directory, so you need to
+rebuild the image by now.
+
+
+1. Install [dracut-iguana](https://github.com/openSUSE/iguana/tree/main/dracut-iguana) from
+[OBS](https://build.opensuse.org/package/show/home:oholecek:iguana/dracut-iguana).
+
+2. (Optional) For debugging purposes, you might want to add SSH support so you can connect to the
+   system. If that's the case, please install the `dracut-ssh` package and place your public SSH key
+   on `/root/.ssh/authorized_keys`.
+
+3. Create a `/etc/NetworkManager` directory to be included in the initrd:
+
+        mkdir -p rd.live.networkmanager/etc/NetworkManager
```

Comment:
Why is this needed?

Reply:
Because we need to bind-mount the `/etc/NetworkManager` directory in the backend container to copy the configuration files at the end of the installation.

Reply:
Ok, this leaves more to explain.
This will create directory in the initrd fs. Once initrd switches root or reboots, then this dir is lost.

Who is the supposed consumer of this configuration? Initrd or installed system?

Reply:
NetworkManager runs in the host system, although you can change the configuration using its D-Bus interface from D-Installer web UI. The changes are written by NetworkManager to that directory.

At the end of the installation, `dinstaller-backend` needs to copy those files to the installed system. If you reboot into the installed system, those files should be there.

Thinking about it twice, I guess we could define a "step" to create the directory in the workflow itself. Is it possible to run an arbitrary command in the host (without using a container)? I could use a container to create the directory, but it is probably too much :)

Reply:
I guess support for [steps](https://github.com/openSUSE/iguana/blob/ba332ca6aa38d607e816e26b9b4a526afbeb21b3/iguana-workflow/examples/control.yaml#L23-L24) is not implemented yet. I am just guessing because the [steps field](https://github.com/openSUSE/iguana/blob/1fbfbe7ebf5d1993e37945bccd0ba1b7f5c46ca3/iguana-workflow/src/workflow.rs#L34) is not referenced.

Reply:
> NetworkManager runs in the host system, although you can change the configuration using its D-Bus interface from D-Installer web UI. The changes are written by NetworkManager to that directory.
> 
> At the end of the installation, `dinstaller-backend` needs to copy those files to the installed system. If you reboot into the installed system, those files should be there.

However iguana is running in initrd. Anything you create at initrd stage in root (like /etc/NM) will not be visible in the booted system and is also not preserved for next boot. Usually system to boot is mounted under `/sysroot` so you would want `/sysroot/etc/NetworkManager`. Iguana mounts it after all containers run based on `/iguana/mountlist` file. I still think this is not doing what you expect it.

Maybe we are talking about different things here so lets go to basics. D-Installer partition disk, format and install root fs. So D-installer container have next root mounted at some point. And you want to copy generated network manager config to this mounted root. Right?

If you just need to have shared storage between backend and web container then you can already used any volume (like `dbus` volume is already shared)

> 
> Thinking about it twice, I guess we could define a "step" to create the directory in the workflow itself. Is it possible to run an arbitrary command in the host (without using a container)? I could use a container to create the directory, but it is probably too much :)

Yes, steps are not implemented (yet, I am not sure if we need/want them). Using different container will not help as container do not have access to initrd fs.

Reply:
> Maybe we are talking about different things here so lets go to basics. D-Installer partition disk, format and install root fs. So D-installer container have next root mounted at some point. And you want to copy generated network manager config to this mounted root. Right?

Yes, I tend to think we are talking about different things 😅. I will try to summarize our POV (that might be wrong):

* NetworkManager runs on initrd (`nm-initrd.service`).
* The user can alter the configuration through the D-Installer UI (`dinstaller-web` service), which talks to NetworkManager via D-Bus. That's why we added `/var/run/dbus/system_bus_socket` to the list of volumes.
* Before finishing the installation, `dinstaller-backend` needs to copy the NetworkManager configuration files (from the initrd), to the installed system. So it needs to access the `/etc/NetworkManager` directory in initrd.

Mounting `/etc/NetworkManager` as a volume would make the trick. The problem is that the /etc/NetworkManager directory does not exist from the very beginning (only after you make a change in the configuration). So we cannot bind-mount the directory in the `dinstaller-backend`.

Does it explain the problem better?

> 
> If you just need to have shared storage between backend and web container then you can already used any volume (like `dbus` volume is already shared)

Not between the backend and the web container, but between the initrd and the backend.

> Yes, steps are not implemented (yet, I am not sure if we need/want them). Using different container will not help as container do not have access to initrd fs.

Unless you make some directory accessible as a volume, right?

Reply:
> * Before finishing the installation, `dinstaller-backend` needs to copy the NetworkManager configuration files (from the initrd), to the installed system. So it needs to access the `/etc/NetworkManager` directory in initrd.

But this will not copy it to the installed system. Unless NetworkManager does some tricks like copying initrd `/etc/NetworkManager` to installed system `/sysroot/etc/NetworkManager` (which is only available after all containers are finished). I checked NM dracut module and I did not find anything doing such thing.

> Not between the backend and the web container, but between the initrd and the backend.

So I guess my question is what then copies config from initrd to installed system.

Reply:
> > * Before finishing the installation, `dinstaller-backend` needs to copy the NetworkManager configuration files (from the initrd), to the installed system. So it needs to access the `/etc/NetworkManager` directory in initrd.
> 
> But this will not copy it to the installed system. Unless NetworkManager does some tricks like copying initrd `/etc/NetworkManager` to installed system `/sysroot/etc/NetworkManager` (which is only available after all containers are finished). I checked NM dracut module and I did not find anything doing such thing.

Perhaps the "installed system" means different things: I am referring to the system that `dinstaller-backend` is installing, which is mounted under `/mnt` in the `dinstaller-backend` container.

> 
> > Not between the backend and the web container, but between the initrd and the backend.
> 
> So I guess my question is what then copies config from initrd to installed system.

The `dinstaller-backend` does it (getting the files from the bind-mounted `/etc/NetworkManager` directory).

I guess moving this discussion offline might help to clarify the situation.

Reply:
After discussing this issue, we decided to implement https://github.com/openSUSE/iguana/issues/18 so rebuilding the initrd is not needed anymore.